requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+- **Terminate-able** - We should be able to rip the carpet out from under the Task
+- **Nonblocking failures** - A failure of one task should not halt the execution of others.
+
+### Historical Exports *(AKA Vanilla ETL)*
+
+We do a lot of moving around data. Actually, a good chunk of our business is just getting data form A to B safe and sound, and in on piece. Some of these jobs are short, some are long running. There are a few requirements and nice to haves around ETLs
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Nice to pick up where you left off
+- **DAG-able** - Most ETLs resemble a Directed Acyclical Graph. Move A → B → C
+    - You can think of this as **Composable**
+- **Parallelize-able** - We should be able to run multiple bits of the DAG at once
+- **Timeouts** - We should be able to set timeouts on the DAG and Nodes
+- **Cancelable** - We should be able to cancel the DAG’s execution *gracefully*
+- **Terminate-able** - We should be able to rip the carpet out from under the DAG
+- **Nonblocking failures** - A failure of one DAG should not halt the execution of others.
+
+### Webhooks
+
+Ahh webhooks. Webhooks are a particularly sticky problem.
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Helps with idempotency/exactly once
+- **Parallelize-able** - We should be able to run multiple webhooks at once and out of order (non-blocking)
+- **Timeouts** - We should be able to set timeouts for webhooks
+- **Terminate-able** - We should be able to rip the carpet out from under webhooks
+- **Nonblocking failures** - A failure of one webhook endpoint should not halt the execution of others.
+
+## Requirements Grid
+
+| Task | Idempotent / Exactly Once | Retry-able | Resumable | Parallelize-able | Timeouts | Cancelable | Terminate-able | Nonblocking Failures |
+| --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| Billing | 💯 | 💯 | 💯 | 🙏 | 🙏 | 🙏 | 💯 | 💯 |
+| Celery Task |  | 💯 | 🙏 | 💯 | 💯 |  | 💯 | 💯 |
+| ETLs | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 |
+| Webhooks | 💯 | 💯 | 🙏 | 💯 | 🙏 |  | 💯 | 💯 |
+
+🙏 - Really nice to have
+
+💯 - Required
+
+# Where are we at now?
+
+Currently our plugin server has been doing all of the heavy lifting and has gotten us quite far in the **************2 years************** that it has existed. I once gave feedback that the plugin server is a diamond encrusted hammer that tempts you into thinking that everything is a nail. Joke was on me because it turned out that it got us way further than I thought.
+
+Since this began we’ve done amazing work in keeping the plugin server going. Chief among the improvements is that we split the plugin server up dependent on the workload into pools of resources that are responsible for different categories of plugin.
```

Comment:
We chose Kubernetes as our workload scheduler and, although I have lots of negative feelings towards k8s, I support using its capabilities instead of rebuilding them ourselves.

We are at a growth stage where incidents will be common (both due to unpredictable input volumes, and self-inflicted bugs), and one of the things we need to prioritize is reducing the blast radius of these incidents. Splitting up resources is a foolproof way to achieve this.

Reply:
I hope we can just agree to disagree here then. I won't stand in anyone's way (like I wrote 2x in my original comment), just wanted to understand the tech better and got many answers.

However, I _do_ think we _absolutely should_ write our own k8s, otherwise we're open for disruption. I'd prefer we're the ones doing the disruption.

Don't take that statement 100% literally - no, we shouldn't build exactly what the Kubernetes Platform is, or even close. And no, we don't have the resources to undertake these things right now.

But we should stop fearing hard engineering challenges! Whether it's a query engine, a mobile session recording toolkit, an orchestration layer or anything in between. Let's instead be the company that has solved these problems and is offering them to our users as part of an integrated solution. 🤷